### DISCUSSION TITLE
type Typevar bound interaction

### DISCUSSION THREAD

**Discussion by hmc-cs-mdrissi** [microsoft/pyright#5580] [Q&A] [answered]
I'm confused by an interaction between bound=type TypeVar vs TypeVar using type[T].

```python
from typing import TypeVar

T = TypeVar("T")
T_type = TypeVar("T_type", bound=type)


def f1(x: type[T]) -> type[T]:
    return x


def f2(x: type) -> type:
    return x


def g(x: T_type) -> T_type:
    print(f1(x)) # Error produced here
    print(f2(x))
    return x

```

The error produced,

```
Argument of type "T_type@g" cannot be assigned to parameter "x" of type "type[T@f1]" in function "f1"
```

I don't follow why though. In g, x is type variable bound to type so it always a type. f1 should accept any type and return it back. f2 adjusts it type unsubscripted and is fine on next line.

Was unsure to open this as bug/question and if there's some missing rule between bound=type vs type[T] I'm not understanding.

edit:

```python
class Foo(Generic[T]):
    ...


T_foo = TypeVar("T_foo", bound=Foo[object])


def h1(x: Foo[T]) -> Foo[T]:
    return x


def h2(x: Foo[Any]) -> Foo[Any]:
    return x


def i(x: T_foo) -> T_foo:
    print(h1(x))
    print(h2(x))
    return x
```

This code with Foo instead of type does pass type checking and is similar in spirit to above but using different generic type then special type. I did Foo[Any] as I think of type equivalent type[Any].

**Comment by erictraut** [comment, marked as answer]
That's a legit bug. I'll create a bug report from this discussion.
